File: scripts/makegradient.pl
#!/usr/bin/perl
#
# File: makegradient.pl
# Written by Benjamin Mampaey on 4 Feb 2010
# This script will generate a gradient

use strict;

my $numberColors = 0;
my $maxNumberColors = 0;
my $maxL = 50;
my $width = 20;
my @S = (100, 50);
my @L = (100, 75, 87.5, 62.5, 93.75, 68.75, 81.25);
my @H = (0, 60, 120, 180, 240, 300, 30, 90, 150, 210, 270, 330, 15, 75, 135, 195, 255, 315, 45, 105, 165, 225, 285, 345); 
my @P = (0, 120, 240, 360);
my $filename = "gradient";
my $c = 1;
my $p = 1;
my $i = 1;
my @magick_colors;
my @dot_colors;
my $preview;

#We read the arguments
use Getopt::Std;
my %opt;
my $opt_string = 'n:l:o:icph';
getopts( "$opt_string", \%opt ) or usage();
if( $opt{h} )
{
	print "Create a gradient of color in hsb format with Imagemagick, or to include in C.\nThe gradient allways start with the black color.\n";
	print "Options :\n";
	print "\t-h : This help and exit.\n";
	print "\t-i : Make a png gradient usable with imagemagick clut option";
	print "\t-c : Make the corresponding c header file.\n";
	print "\t-o : Filename for output.\n";
	print "\t-n : Number of colors in the gradient (all of them if 0 or not specified).\n";
	print "\t-l : Maximal lightness (between 50 and 100) for primary colors (red, blue, green). 100 if you want primary colors.\n";
	print "\t-p : Generate a preview (gradientpreview.png).\n";
	
	exit 0;
}
if( $opt{n} )
{
	$maxNumberColors = $opt{n};
}
if( $opt{l} )
{
	$maxL = $opt{l};
}
if( $opt{c} || $opt{i} || $opt{p})
{
	$c = $opt{c} ? 1: 0;
	$i = $opt{i} ? 1: 0; 
	$p = $opt{p} ? 1: 0;
}
if( $opt{o} )
{
	$filename = $opt{o};
}


# We create the black

push @dot_colors, '[color=\"0, 0, 0\"];';
push @magick_colors, 'hsb(0%, 0%, 0%)';



# We add the colors

COLOR : while (1)
{
	foreach my $s (@S)
	{
		foreach my $l (@L)
		{
			HUE : foreach my $h (@H)
			{
				foreach my $p (@P)
				{
	 
					if ($h > $p - $width && $h <= $p && $l > (((100 - $maxL)/$width) * ($h - $p)) + $maxL) #Too close of a primary color
					{
						next HUE;
					}
					if ($h < $p + $width && $h > $p && $l > (((100 - $maxL)/$width) * ($p - $h)) + $maxL) #Too close of a primary color
					{
						next HUE;
					}
				}

				push @dot_colors, '[color=\"'.sprintf("%.2f", ($h/360)) . ", " . sprintf("%.2f", ($s/100)) . ", " . sprintf("%.2f", ($l/100)) . '\"];';
				push @magick_colors, "hsb(".sprintf("%.2f", ((100*$h)/360))."%, $s%, $l%)";
				++$numberColors;
				last COLOR if($maxNumberColors != 0 && $numberColors >= $maxNumberColors);
						
			}
		}
	}
	
	last COLOR if($maxNumberColors == 0);
}


if($c)
{
	my $magick_string = join '", "', @magick_colors;
	$magick_string = '"'.$magick_string.'"';
	my $dot_string = join '", "', @dot_colors;
	$dot_string = '"'.$dot_string.'"';
	my $cpp_filename = $filename.".cpp";
	open FILE, ">$cpp_filename" or die "Could'nt create $cpp_filename : $!";
	print FILE "#include \"gradient.h\"\n";
	print FILE "const unsigned gradientMax = $numberColors;\n";
	print FILE "const char* dot_gradient[] = {$dot_string};\n";
	print FILE "const char* magick_gradient[] = {$magick_string};\n";
	close FILE; 
	my $h_filename = $filename.".h";
	open FILE, ">$h_filename" or die "Could'nt create $h_filename : $!";
	print FILE "#ifndef GRADIENT_H\n";
	print FILE "#define GRADIENT_H\n";
	print FILE "extern const unsigned gradientMax;\n";
	print FILE "extern const char* dot_gradient[];\n";
	print FILE "extern const char* magick_gradient[];\n";
	print FILE "#endif\n";
	close FILE; 

}
if($i) #ImageMagick
{
	my $magick_string = join '" "xc:', @magick_colors;
	$magick_string = '"xc:' . $magick_string . '"';
	my $i_filename = $filename.".png";
	print STDERR `convert -size 1x1 $magick_string -append $i_filename`."\n";
}
if($p)
{
	my $magick_string = shift @magick_colors;
	$magick_string = ' -size 150x20 -background "'.$magick_string.'" -gravity center -fill white -pointsize 12 label:"'.$magick_string.'"';
	$magick_string.= ' -size 150x20 -background "'.$_.'" -gravity center -fill black -pointsize 12 label:"'.$_.'"' foreach(@magick_colors);
	my $p_filename = $filename."_preview.png";
	print STDERR `convert $magick_string -append $p_filename`."\n";
}
print STDERR "Created a gradient of $numberColors colors.\n";

File: scripts/make_gradient.py
import math
import random
import Image, ImageDraw
import logging
import sys
import argparse

def distance(a, b):
	
	result = 0
	for x in zip(a, b):
		result += (x[0] - x[1])**2
	
	return math.sqrt(result)

def grey_level(color):
	return 100. - (((255.*(color[2]-color[1]))**2 + (255.*(color[0]-color[2]))**2 + (255. *(color[1]-color[0]))**2)/84565012.5)

def draw_gradient(colors, filename):
	image = Image.new("RGB", (1, len(colors)))
	draw = ImageDraw.Draw(image)
	for y in range(len(colors)):
		draw.rectangle([(0, y),(1, y+1)], fill=colors[y])
	image.save(filename)

def draw_gradient_preview(colors, filename):
	width = 100
	height = 20
	image = Image.new("RGB", (width, len(colors)*height))
	draw = ImageDraw.Draw(image)
	for y in range(len(colors)):
		draw.rectangle([(0, y * height),(width, (y+1) * height)], fill=colors[y])
		draw.text((1, (y+0.1) * height), str(colors[y]), fill=(128,128,128))
	image.save(filename)

def write_C_header(filename, colors):
		web_colors = ['"#%02x%02x%02x"' % color for color in colors]
		try:
			with open(filename, 'w+') as f:
				f.write("#ifndef GRADIENT_H\n")
				f.write("#define GRADIENT_H\n")
				f.write("const unsigned gradientMax = "+str(len(colors))+";\n")
				f.write("const char* const gradient[] = {"+','.join(web_colors)+"};\n")
				f.write("#endif\n")
		except IOError:
			log.critical("Error writing to file " + str(filename)+ ": " + str(why))

def write_Python_header(filename, colors):
		web_colors = ['"#%02x%02x%02x"' % color for color in colors]
		try:
			with open(filename, 'w+') as f:
				f.write("gradient = ["+','.join(web_colors)+"]\n")
		except IOError:
			log.critical("Error writing to file " + str(filename)+ ": " + str(why))


# Start point of the script
if __name__ == "__main__":
	# Get the arguments
	parser = argparse.ArgumentParser(description='Generate a gradient of color as a png image, a c include file, a python list of colors.')
	parser.add_argument('--verbose', '-v', default=False, action='store_true', help='Verbose')
	parser.add_argument('--increment', '-i', default=30, type=int, help='The minimal increment between colors. Must be a value between 1 and 255')
	parser.add_argument('--max_grey_level', '-g', default=80, type=int, help='The maximal accepeted grey level in percent. Must be a value between 0 and 100')
	parser.add_argument('--window_size', '-w', default=10, type=int, help='The size of the running window in wich colors must not be similar ')
	args = parser.parse_args()
	
	console = logging.StreamHandler()
	console.setFormatter(logging.Formatter('%(levelname)-8s: %(message)s'))
	if args.verbose:
		console.setLevel(logging.INFO)
	else:
		console.setLevel(logging.CRITICAL)
	
	# Create logger
	log = logging.getLogger('make_gradient')
	log.addHandler(console)
	
	# set up of the global variables
	if args.increment < 1 or args.increment > 255:
		log.critical("increment must be between 1 and 255, has been set to " + str(increment))
		sys.exit(2)
	else:
		increment = args.increment
	
	if args.max_grey_level < 0 or args.max_grey_level > 100:
		log.critical("max_grey_level must be between 0 and 100, has been set to " + str(increment))
		sys.exit(2)
	else:
		max_grey_level = args.max_grey_level
	if args.window_size < 0:
		log.info("window_size must not be smaller than 0, forcing to 0")
		window_size = 0
	else:
		window_size = args.window_size
	
	colors = list()
	# We generate all the non greyish possible colors
	for red in range(0,256,increment):
		for green in range(0,256,increment):
			for blue in range(0,256,increment):
				color = (red,green,blue)
				if grey_level(color) < max_grey_level:
					colors.append(color)
				else:
					log.info("Discarding greyish color " + str(color))

	# We separate the similar colors as much as we can in a running window
	random.shuffle(colors)
	separated_colors = [colors.pop()]
	while len(colors) > 0:
		furthest = None
		max_distance = 0
		for c in colors:
			dist = min([distance(c, s) for s in separated_colors[-window_size:]])
			if dist > max_distance:
				max_distance = dist
				furthest = c
		color = furthest
		colors.remove(color)
		separated_colors.append(color)

	colors = separated_colors
	
	draw_gradient(colors, "gradient.png")
	draw_gradient_preview(colors, "gradient_preview.png")
	
	# We write the C header
	write_C_header("gradient.h", colors)
	# We write the python header
	write_Python_header("gradient.py", colors)

